ethereum/eth_watcher.go

```diff
@@ -134,14 +133,41 @@ func (e *ETHWatcher) Receive() <-chan Balance {
 	return e.balanceChan
 }
 
-func (e *ETHWatcher) updateBalances() error {
+func (e *ETHWatcher) updateBalances() {
 	e.addressToBalanceMu.Lock()
+	e.addressToBalanceMu.Unlock()
 	addresses := []common.Address{}
 	for address := range e.addressToBalance {
 		addresses = append(addresses, address)
 	}
-	e.addressToBalanceMu.Unlock()
+	// Intentionally ignore addresses we failed to fetch balances for
+	// and simply attempt them again at the next polling interval
+	addressToBalance, _ := e.getBalances(addresses)
+	for address, newBalance := range addressToBalance {
+		if cachedBalance, ok := e.addressToBalance[address]; ok {
+			if cachedBalance.Cmp(newBalance) != 0 {
+				e.addressToBalance[address] = newBalance
+				updatedBalance := Balance{
```

Comment:
nit: The names `updatedBalance` and `newBalance` seem like synonyms to me but they have different purposes here. The fact that the `Balance` type has a field called `Balance` is also sort of confusing.

What do think about this naming scheme instead?

```go
type Balance struct {
    Address common.Address
    Amount *big.Int
}
```

The idea here is that an `Amount` by itself is just a number. We only call it a `Balance` when it is associated with an account.

This would lend clarity to the code I mentioned by changing it to:

```go
for address, newAmount := range addressToAmount {
    // ...
    updatedBalance := Balance{
        Address: address,
        Amount: newAmount,
    }
}
```

What do you think?

Reply:
I like it! Much clearer. Thanks for the suggestion.